### PR TITLE
Initial draft of communication guide

### PR DESCRIPTION
**Pull request by snoyberg**

### CODE REVIEW

File: COMMUNICATIONS.md [hunk 1]
```diff
@@ -1,3 +1,198 @@
...
+
+The following are principles I recommend. If you disagree with these,
+would like to propose changes to the wording, or would like to propose
+others, feel free to open an issue or pull request on this repository.
+
+### Don't take it personally
+
+We often become personally invested in our software. We dedicate huge
+amounts of time, both personal and professional, to making it
+succeed. It can sometimes be difficult to hear criticism of this
```
Comment (vedksah) [inline COMMUNICATIONS.md]:
```suggestion
succeed. It can sometimes be difficult to hear criticism of our
```

File: COMMUNICATIONS.md [hunk 2]
```diff
@@ -1,3 +1,198 @@
...
+proposal](https://www.snoyman.com/blog/2018/11/proposal-stack-coc)
+mentions, this in itself is not ideal, since it is somewhat
+divisive. Allow me to backtrack that divisiveness here a bit: we
+should _not_ be trying to establish a separate Stack community that
+does not integrate with other projects.
+
+Much of how Stack has been designed is in fact along these lines:
+
+* The Stackage project is separate, and is usable by both Nix and (to
+  some extent) cabal-install
```
Comment (vedksah) [inline COMMUNICATIONS.md]:
```suggestion
  some extent) cabal-install (the tool)
```

File: COMMUNICATIONS.md [hunk 3]
```diff
@@ -1,3 +1,198 @@
...
+* Hackage remains the upstream repository
+* Cabal remains the underlying build system
```
Comment (vedksah) [inline COMMUNICATIONS.md]:
```suggestion
* Cabal (the library) remains the underlying build system
```

**Review by vedksah — APPROVED** [review]
This looks great! Got some minor suggestions.

It's certainly not the Stack project that needs a code of conduct the most but setting a good example will hopefully inspire others to follow suit!

**Comment by snoyberg** [comment]
@vedksah I appreciate the suggestions here, they're helpful. However, I'd like to request that you avoid implications against other projects. This repo is about improving the Stack community, there's no need to reference other projects, even via implication.

File: COMMUNICATIONS.md (hunk 1 above)
Comment (snoyberg) [reply COMMUNICATIONS.md]: Agreed. Also, this is the first new suggestion usage I've seen. I'll include in a commit I apply locally.

File: COMMUNICATIONS.md (hunk 2 above)
Comment (snoyberg) [reply COMMUNICATIONS.md]: Agreed

File: COMMUNICATIONS.md (hunk 3 above)
Comment (snoyberg) [reply COMMUNICATIONS.md]: Actually, this one isn't quite right. I'm referring to Cabal as the build system standard here, which is slightly different from Cabal the library. I think I'll include reference to both.

**Comment by vedksah** [comment]
Ok, got it! I already inadvertently failed to honour the code of conduct :laughing:

**Comment by snoyberg** [comment]
No problem, we're all learning. The most important thing is doing better next time.

File: COMMUNICATIONS.md [hunk 4]
```diff
@@ -1,3 +1,199 @@
...
+forms:
+
+* **BAD** Their previous work is shoddy, so this work must be shoddy.
+* **BAD** They are part of an organization/company/committee,
+  therefore they are inherently biased and cannot be listened to.
+* **BAD** They are seen as friendly with someone else, therefore they
+  are inherently biased.
+
+A related topic is the pattern of behavior argument. "Person X has
+done A, B, and C, therefore D must have a similar motivation." That's
```
Comment (duplode) [inline COMMUNICATIONS.md]:
```suggestion
done A, B, and C, therefore Y must have a similar motivation." That's
```

File: COMMUNICATIONS.md [hunk 5]
```diff
@@ -1,3 +1,199 @@
...
+project. However, the definition of technical merits is in fact quite
+broad, including:
+
+* Performance
+* Reliability
+* Maintainability
+* Ease of use
+* Marketability
+
+I believe that last one may be a surprise to many. I want to make the
```
Comment (duplode) [inline COMMUNICATIONS.md]:
```suggestion
I believe the last one may be a surprise to many, so I want to make the
```

File: COMMUNICATIONS.md [hunk 6]
```diff
@@ -1,3 +1,199 @@
...
+divisive. Allow me to backtrack that divisiveness here a bit: we
+should _not_ be trying to establish a separate Stack community that
+does not integrate with other projects.
+
+Much of how Stack has been designed is in fact along these lines:
+
+* The Stackage project is separate, and is usable by both Nix and (to
+  some extent) cabal-install (the build tool)
+* Hackage remains the upstream repository
+* Cabal (the library, file format, and build system) remain the
```
Comment (duplode) [inline COMMUNICATIONS.md]:
```suggestion
* Cabal (the library, file format, and build system) remains the
```

File: COMMUNICATIONS.md [hunk 7]
```diff
@@ -1,3 +1,199 @@
...
+  underlying mechanism for representing and building packages
+
+Broadening this a bit: I (Michael) am one of the original authors of
+Stack. I'm also the original author of Yesod and Conduit. It is
+absolutely vital that Stack treat other web frameworks (like Servant)
```
Comment (duplode) [inline COMMUNICATIONS.md]:
```suggestion
absolutely vital that Stack treats other web frameworks (like Servant)
```

File: COMMUNICATIONS.md [hunk 8]
```diff
@@ -1,3 +1,199 @@
...
+
+This is covered in depth in the "prior art" links above. Most people
+are interacting with the Stack community in their free time. Companies
+that support development of tooling are spending funds they could be
+using for non-community activities. We should respect these
+contributions.
+
+On the other hand, for users and consumers of the technology, we
+should wherever possible try to make life as easy as we can. Avoid
+blocking people if you can. Recognize differences in goals and see if
```
Comment (duplode) [inline COMMUNICATIONS.md]:
```suggestion
creating blockers for people if you can. Recognize differences in goals and see if
```

Comment (duplode) [reply COMMUNICATIONS.md]: Phrasing -- "Avoid blocking people" brings to mind "blocking as in Twitter", as opposed to "creating a technical obstacle".

File: COMMUNICATIONS.md [hunk 9]
```diff
@@ -1,3 +1,199 @@
...
+
+Keep discussions on topic. If there's an issue about X, do not
+unnecessarily raise Y. Do not rehash previous arguments. Do not open
+copious issues to rehash previous arguments. If there is an inherent
+disagreement, accept that that disagreement exists, and do not attempt
+to force others to comply with your viewpoint.
+
+### Don't demand changes from others
+
+These guidelines are non binding. No one has the right to point at
```
Comment (duplode) [inline COMMUNICATIONS.md]:
```suggestion
These guidelines are non-binding. No one has the right to point at
```

File: COMMUNICATIONS.md [hunk 10]
```diff
@@ -1,3 +1,199 @@
...
+environment you are not comfortable. If a discussion has been handled
+in a manner which you're not happy with, feel free to either disengage
+completely, or explain that you're stepping away, possibly with a link
+to this section.
+
+If you feel comfortable with doing so, in some cases a private message
+to the other party can help resolve a situation you're unhappy
+with. They may be unaware that you're uncomfortable. They may be more
+sensitive to your feelings if expressed privately and calmly. Again,
+you may not _demand_ such a change, but a polite request is
```
Comment (duplode) [inline COMMUNICATIONS.md]:
```suggestion
you cannot _demand_ such a change, but a polite request is
```

**Review by duplode — CHANGES_REQUESTED** [review]
Minor phrasing tweaks and typo fixes.

File: COMMUNICATIONS.md (hunk 4 above)
Comment (snoyberg) [reply COMMUNICATIONS.md]: I actually meant something slightly different, I'll include a change to see if it clarifies.

File: COMMUNICATIONS.md (hunk 5 above)
Comment (snoyberg) [reply COMMUNICATIONS.md]: 👍

File: COMMUNICATIONS.md (hunk 6 above)
Comment (snoyberg) [reply COMMUNICATIONS.md]: 👍

File: COMMUNICATIONS.md (hunk 7 above)
Comment (snoyberg) [reply COMMUNICATIONS.md]: 👍

File: COMMUNICATIONS.md (hunk 8 above)
Comment (snoyberg) [reply COMMUNICATIONS.md]: Ooh, really nice catch!

File: COMMUNICATIONS.md (hunk 4 above)
Comment (snoyberg) [reply COMMUNICATIONS.md]: I ended up with this, how's that?

```diff
-done A, B, and C, therefore D must have a similar motivation." That's
+done A, B, and C, therefore when X does D it must have a similar motivation." That's
```

**Comment by dbaynard** [comment]
I suggest moving 'Don't take it personally' down the list.

While it is a very important piece of advice — that we have to try to approach our work objectively — it may be better to begin with advice on giving, rather than receiving feedback. As it is, it read to me like 'the most important thing is to receive the message well'. Whoever sends a message has the first responsibility to ensure it follows the guidelines; whoever receives that message only then takes the responsibility to listen to the content. Changing the order matches the responsibilities.

**Comment by dbaynard** [comment]
One thing currently missing (though I may have missed it) is how to deal with, and possibly resolve, conflict. The proposed code of conduct defers to the project leads. The communication guide might set expectations of how this would work.

**Comment by snoyberg** [comment]
I'm fine with moving the paragraph lower. I'll be honest: I wrote down a bunch of headings as they came to me, and didn't think about their relative position. There may be even more reorderings that are worth having.

See issue #1 for a discussion of enforcement.

**Comment by gbaz** [comment]
What is the purpose of a separate communications guide?

The contributor covenant (which is very widely used, including by linux, and is suggested in the other PR) seems like it suffices to cover communications, as it lists the following:

* Using welcoming and inclusive language
* Being respectful of differing viewpoints and experiences
* Gracefully accepting constructive criticism
* Focusing on what is best for the community
* Showing empathy towards other community members

 I don't see what purpose a separate communications guide holds, especially since the current draft, if anything, seems to cut against the spirit of the above, and be weaker. In particular, the draft communications guide says  "These guidelines are non-binding." But the CoCov has a section on "Enforcement." Further, the draft communications guide says "don't take it personal" -- but this is advice to someone that feels that they are on the receiving end of a personal attack. Since personal attacks are actually ruled out by the CoCov to begin with, which has an "Enforcement" clause, why is this guide telling people to ignore them? If anything, wouldn't "don't make it personal" suffice?

Anyway, the more full take I have is that the CoCov is widely adopted and essentially self-sufficient. Nearly all projects that have adopted it have not felt the need to augment it with things, especially not things that may run against it in some ways. To me, adopting it, and then adopting a whole other thing as well, feels like adopting a standard open source license and then tossing in a whole bunch of other custom clauses, and I'd suggest not doing so.

The one thing I can see that is particular here is the section on "avoid partisanship" but I don't think this belongs in a CoC at all -- rather, it is a general statement about the purpose of stack, and could well live in stack-specific documentation, because it pertains to the goals of stack, not to how to conduct discussions relevant to these goals.

Finally, I also question the "respect privacy" section since it seems to again introduce the principle only to immediately introduce language that undermines it: "As such, communications can only reasonably
be considered private if both sides agree." But how can both sides agree when one side sends the email and the other receives it?  All this says is "respect privacy except when you feel you have a reason not to," at which point, you might as well say nothing at all.

**Comment by snoyberg** [comment]
You're free to disagree with me and my motivations. However, I have given my full explanation for why I want this separate document in my blog post, so I have nothing to add here.

**Comment by gbaz** [comment]
Well there are both general and specific concerns that I don't think the blog post addresses. I'm not imputing any motivations nor disagreeing with them.

I'm suggesting a technical problem -- two documents, which potentially conflict, don't make for a good joint CoC. Similarly two licenses which both have different clauses covering the same things, don't make for a clear single license. What I am suggesting is that whatever CoC is adopted be authoritative, and any further guide have its status clarified as a set of personal notes as to the interpretation and spirit of things.

I'm also proposing that the "avoid partisanship" section be moved just to an entirely distinct thing that's part of the mission statement of stack.

And finally, I'm suggesting that the "respect privacy" section does not make sense, and similarly the "don't take it personal" section. In the latter, the point that one should distinguish personal criticism from technical discussion can fit perfectly well under the subhead "don't make it personal" much more clearly.

**Comment by dbaynard** [comment]
> What is the purpose of a separate communications guide?

For me, the value is that communications and conduct are not quite the same thing. The covenant does not have anything to say about the project, and there is likely to be best practice in communicating that doesn't relate to the pleasantness of the community.

The covenant is also quite abstract; there may be value in a more detailed document — as long as that document complements it.

>  the draft communications guide says "don't take it personal" -- but this is advice to someone that feels that they are on the receiving end of a personal attack.

That's a good point. I agree with the sentiment (a mix of 'gracefully accept constructive criticism' and an attempt to avoid escalation) but it ought to say that explicitly.

Edit: Should have hit refresh before submitting the comment!

File: COMMUNICATIONS.md (hunk 4 above)
Comment (duplode) [reply COMMUNICATIONS.md]: Crystal clear. I had misunderstood it indeed; this dispels my confusion.

**Comment by DanBurton** [comment]
The distinction between the two documents could be that the CoC is enforced, while the communication guide is merely suggested. That way, if any contradictions are detected, the CoC generally trumps the communication guide. (However, the two documents are, afaict, reasonably harmonious).

**Comment by snoyberg** [comment]
@dbaynard You've captured my thoughts on the different documents perfectly, and expressed them better than I could. Thank you.

With the "don't take it personally," I'm actually getting at something very specific, and perhaps I need to be more explicit about it. I want to encourage people (and, more honestly, myself!) to try and decouple "criticism of my code" and "criticism of me as a person." There's certainly value in differentiating between criticism and attack. If I'm told that I'm not an organized person, that's criticism, but not necessarily an attack. But I'm not even getting to that level. I'd like to encourage the thought that it's healthy to separate the code we write from our own personal identities.

Maybe this is a problem I've had, and others don't have as much. But remember, these are my personal guidelines, and I'm trying to capture things that have affected me.

**Comment by DanBurton** [comment]
I'm not sure I see the purpose of the communications guide, specifically, why it is paired in the same repo as a code of conduct. This relates to my uncertainty over why these documents are deemed "unofficial." I would suggest moving in one of two directions:

* Make the CoC official, and turn the communications guide into a personal blog post
* Make the CoC official, and make the communications guide an official "addendum" to it

By "make official", I would actually recommend this be reparented under `commercialhaskell` and considered organization-wide for all commercialhaskell projects, forums, and repos (not just stack-related stuff).

**Comment by snoyberg** [comment]
Again, I have no objection to adding a CoC to Stack, or commercialhaskell generally. I feel less good about the communication guide going there. This is very much guidelines, and I don't feel like giving them any air of authority, because they have none.

**Comment by dbaynard** [comment]
@simonmichael made a very constructive comment on reddit, pointing out the [GNU Kind Communications Guidelines](https://www.gnu.org/philosophy/kind-communication.html). See [email introduction](https://lists.gnu.org/archive/html/info-gnu/2018-10/msg00001.html) and [Hacker News discussion](https://news.ycombinator.com/item?id=18274663).

It is well worth reading. The needs are slightly different to stack, but the broad pattern holds.

One interesting aspect is the choice of language — the [anaphora](https://en.wikipedia.org/wiki/Anaphora_(rhetoric)) of beginning each line with 'Please' stands out, and really sells the idea that you as a reader are following the policy because you are a Good Person.

**Comment by snoyberg** [comment]
@dbaynard Thanks for the link. I've looked through the guidelines, and I like what I see there. Providing a link to those guidelines from the COMMUNICATIONS.md document seems fine, after I review it a bit more thoroughly.

But to be clear: this looks to be as opposed to a CoC, not a different choice of CoC. I understand (at least some of) the arguments in that direction, and think the Reddit thread had a really healthy discussion around it. But I wanted to confirm if that's your reading as well.

**Comment by dbaynard** [comment]
I thought the GNU document was in line with these communications guidelines, and somewhat orthogonal to the CoC (which is why I put it here and not on #1). Do note: orthogonal, not opposed. To me, it says much of what you say in the communications guide: in less detail but with a style that addresses some of the issues raised in this discussion.

**Comment by ntc2** [comment]
> Further, the draft communications guide says "don't take it personal" -- but this is advice to someone that feels that they are on the receiving end of a personal attack. Since personal attacks are actually ruled out by the CoCov to begin with, which has an "Enforcement" clause, why is this guide telling people to ignore them? If anything, wouldn't "don't make it personal" suffice?

Personally, I think an explicit "don't take it personal" guideline is a great idea. Let me try to explain why. 

Regarding "Since personal attacks are actually ruled out by the CoCov to begin with, which has an 'Enforcement' clause, why is this guide telling people to ignore them?", my position is that something being a "personal attack" is not easily determined or well defined in general. The receiver might feel like something is a personal attack, while the giver might not intend it that way. You don't know someones **intent**, you only know their **impact** on you. And you have some control over that impact.

Communication is often ambiguous, especially online in text where we don't have voice and body language cues. So, related to "don't take it personal" could be "assume goodwill" or "assume you're dealing with a reasonable person". I.e., given several possible interpretations of a person's intent, that are consistent with their actions, assume the most charitable.

Another way to say this is that *emotional reasoning* is not logically sound: just because you feel bad doesn't mean someone has done something bad to you; you might just feel bad. And, you might be able to think about the situation a different way and not feel bad. And you're in control of that, and you don't need other people to change to achieve that. The "don't take it personal" is a reminder to consider these possibilities, and make yourself more robust communicating in an environment you don't control.

**Comment by dbaynard** [comment]
I like that — changing from a negative 'don't take it personally' to a positive 'assume goodwill'. The GNU guidelines begin with 'Please assume other participants are posting in good faith.'

**Comment by magthe** [comment]
@snoyberg 
 
> But to be clear: this looks to be as opposed to a CoC, not a different choice of CoC. I understand (at least some of) the arguments in that direction, and think the Reddit thread had a really healthy discussion around it. But I wanted to confirm if that's your reading as well.

The way I understood it, the GNU document is instead of a CoC _for GNU_. I agree with @dbaynard that it's (at least fairly and in practical terms) orthogonal to a CoC and can serve as a nice example and inspiration for the Recommended Communications Guide.

**Comment by snoyberg** [comment]
While technically orthogonal, I noticed the following in the email from RMS:

> Some maintainers advocated adopting a "code of conduct" with strict
rules.  Some other free software projects have done this, generating
some resistance.3  Several GNU package maintainers responded that they
would quit immediately.  I myself did not like the punitive spirit of
that approach, and decided against it.

So I believe we have two different issues at play here:

1. From the GNU guidelines themselves: are there aspects we wish to adapt in the communications guidelines we end up with here? That would be highly relevant to this PR.
2. Are we at all interested in the content in the email and Hacker News discussions commenting on CoCs in general? That would be more relevant to #4.

It may have only been me who was confused about potentially discussing the second point here, so I'll leave it be. Coming back to (1): I will review a bit more thoroughly right now and, unless I see a problem, I'll add a commit linking to the GNU guidelines in this PR.

**Comment by snoyberg** [comment]
Thanks for the feedback all. I'm going to merge this PR. Further discussions can happen with additional issues and PRs.

**Comment by duplode** [comment]
Great. I have just opened #7 as a reply of sorts to @gbaz 's concerns about the "Respect privacy" section.

**Pull request merged**